mantaray/received.py

```diff
@@ -316,14 +335,35 @@ def _handle_cap(server_view: views.ServerView, args: list[str]) -> None:
     subcommand = args[1]
     if subcommand == "ACK":
         acknowledged = set(args[-1].split())
+
         if "sasl" in acknowledged:
             server_view.core.send("AUTHENTICATE PLAIN")
+
+        for capability in acknowledged:
+            server_view.core.cap_list.add(capability)
+
     elif subcommand == "NAK":
         rejected = set(args[-1].split())
         if "sasl" in rejected:
             # TODO: this good?
             raise ValueError("The server does not support SASL.")
 
+    else:
+        server_view.core.send("CAP END")
+        raise ValueError("Invalid CAP response. Aborting Capability Negotiation.")
+
+    server_view.core.pending_cap_count -= 1
+
+    # If CAP END is sent before the SASL authentication message (_handle_authenticate()),
+    # the server will respond with 904 SASL authentication aborted.
```

Comment:
These two lines feel to me as a pile of confusing details. How about this, which IMO better expresses what you're really trying to say?

```suggestion
    # If we use SASL, we can't send CAP END until all SASL stuff is done.
```